core/src/main/scala/kafka/server/ReplicaAlterLogDirsThread.scala

```diff
@@ -0,0 +1,259 @@
+
+    // Append the leader's messages to the log
+    partition.appendRecordsToFutureReplica(records)
+    futureReplica.highWatermark = new LogOffsetMetadata(partitionData.highWatermark)
+    futureReplica.maybeIncrementLogStartOffset(partitionData.logStartOffset)
+
+    if (partition.maybeReplaceCurrentWithFutureReplica())
+      removePartitions(Set(topicPartition))
+
+    quota.record(records.sizeInBytes)
+  }
+
+  def handleOffsetOutOfRange(topicPartition: TopicPartition): Long = {
+    val futureReplica = replicaMgr.getReplicaOrException(topicPartition, Request.FutureLocalReplicaId)
+    val currentReplica = replicaMgr.getReplicaOrException(topicPartition)
+    val partition = replicaMgr.getPartition(topicPartition).get
+    val logEndOffset: Long = currentReplica.logEndOffset.messageOffset
+
+    if (logEndOffset < futureReplica.logEndOffset.messageOffset) {
+      warn("Future replica for partition %s reset its fetch offset from %d to current replica's latest offset %d"
+        .format(topicPartition, futureReplica.logEndOffset.messageOffset, logEndOffset))
+      partition.truncateTo(logEndOffset, isFuture = true)
+      logEndOffset
+    } else {
+      val currentReplicaStartOffset: Long = currentReplica.logStartOffset
+      warn("Future replica for partition %s reset its fetch offset from %d to current replica's start offset %d"
+        .format(topicPartition, futureReplica.logEndOffset.messageOffset, currentReplicaStartOffset))
+      val offsetToFetch = Math.max(currentReplicaStartOffset, futureReplica.logEndOffset.messageOffset)
+      // Only truncate the log when current replica's log start offset is greater than future replica's log end offset.
+      if (currentReplicaStartOffset > futureReplica.logEndOffset.messageOffset)
+        partition.truncateFullyAndStartAt(currentReplicaStartOffset, isFuture = true)
+      offsetToFetch
+    }
+  }
+
+  def handlePartitionsWithErrors(partitions: Iterable[TopicPartition]) {
+    if (partitions.nonEmpty)
+      delayPartitions(partitions, brokerConfig.replicaFetchBackoffMs.toLong)
+  }
+
+  def buildLeaderEpochRequest(allPartitions: Seq[(TopicPartition, PartitionFetchState)]): ResultWithPartitions[Map[TopicPartition, Int]] = {
+    val partitionEpochOpts = allPartitions
+      .filter { case (_, state) => state.isTruncatingLog }
+      .map { case (tp, _) => tp -> epochCacheOpt(tp) }.toMap
+
+    val (partitionsWithEpoch, partitionsWithoutEpoch) = partitionEpochOpts.partition { case (tp, epochCacheOpt) => epochCacheOpt.nonEmpty }
+
+    val result = partitionsWithEpoch.map { case (tp, epochCacheOpt) => tp -> epochCacheOpt.get.latestEpoch() }
+    ResultWithPartitions(result, partitionsWithoutEpoch.keys.toSet)
+  }
+
+  def fetchEpochsFromLeader(partitions: Map[TopicPartition, Int]): Map[TopicPartition, EpochEndOffset] = {
+    partitions.map { case (tp, epoch) =>
+      tp -> new EpochEndOffset(Errors.NONE, replicaMgr.getReplicaOrException(tp).epochs.get.endOffsetFor(epoch))
```

Comment:
Sorry, I missed the exception here.

Instead of returning ResultWithPartitions, how about we keep it consistent with ReplicaFetcherThread.fetchEpochsFromLeader(), which returns `tp -> new EpochEndOffset(Errors.forException(t), UNDEFINED_EPOCH_OFFSET)` in the resulting map if the current replica for this topic partition is offline?

Reply:
I have updated the patch to return `tp -> new EpochEndOffset(Errors.forException(t), UNDEFINED_EPOCH_OFFSET)` if the replica is offline. HW will be used to truncate the future replica before the future replica is removed from ReplicaAlterDirThread later.

Reply:
Hmm, do we need to truncate based on HW for those partitions that are already marked as offline? It seems that it's simpler to just let handlePartitionsWithErrors() deals with them before they are removed from ReplicaAlterDirThread.

Reply:
@junrao Strictly speaking, we don't have to truncate the future replica is the current replica is offline. But it is probably more intuitive to return `EpochEndOffset(Errors.forException(t), UNDEFINED_EPOCH_OFFSET)` here because the purpose of this method is `fetchEpochsFromLeader()`. The caller of this method, in this case AbstractFetcherThread.maybeTruncate(), should decide which to handlePartitionsWithError or truncate the replica. Does this make sense?

Reply:
Ok, this is fine then.